HeterogeneousCore/CUDACore/interface/CUDAContextState.h

```diff
@@ -25,28 +25,28 @@ class CUDAContextState {
   friend class CUDAScopedContextProduce;
   friend class CUDAScopedContextTask;
 
-  void set(int device, std::shared_ptr<cuda::stream_t<>> stream) {
+  void set(int device, cudautils::SharedStreamPtr stream) {
     throwIfStream();
     device_ = device;
     stream_ = std::move(stream);
   }
 
   int device() const { return device_; }
 
-  std::shared_ptr<cuda::stream_t<>>& streamPtr() {
+  const cudautils::SharedStreamPtr& streamPtr() const {
     throwIfNoStream();
     return stream_;
   }
 
-  const std::shared_ptr<cuda::stream_t<>>& streamPtr() const {
+  cudautils::SharedStreamPtr releaseStreamPtr() {
     throwIfNoStream();
-    return stream_;
+    return std::move(stream_);
```

Comment:
does the `std::move` make sense here ?
the return is by value, not by rvalue reference.

Reply:
It's either `move` or something along
```cpp
auto tmp = stream_;
stream_.reset();
return tmp;
```
The aim is that `stream_` must be empty after the return (its state is used as a run-time sanity check).

Reply:
Understood, thanks.
Could you add a comment about it, to make sure we don't remove it later on by mistake ?

Would it work to return a `cudautils::SharedStreamPtr &&`, or that would be a reference to a temporary ?

Reply:
I'll add a comment.

> Would it work to return a `cudautils::SharedStreamPtr &&`, or that would be a reference to a temporary ?

IIUC it would likely work, but I'd say returning a temporary is safer. The object is anyway propagated to a member initialization in the constructor of `CUDAScopedContextBase`. Therefore I'd expect copy elision to ensure that only one `move` is done (and in case more is done, this piece is should not be performance-critical).